### PR TITLE
Client-gen: handle dotted group name, e.g., "authentication.k8s.io"

### PR DESCRIPTION
**Pull request by caesarxuchao**
The client-gen used to assume the group name doesn't include dot, but it's not true, e.g., we have group `authentication.k8s.io`.

With this PR, Client-gen will use the full group name when creating directory (e.g., the client for the authentication group will be generated at pkg/client/clientset_generated/release_1_3/typed/`authentication.k8s.io`/v1/). However, because golang doesn't allow dot in variable/function names, so when the group name is used as part of variable/function name, client-gen extracts the part before the first dot (e.g., authentication).

This PR also changes the group name of the test group from `testgroup` to `testgroup.k8s.io` to verify if client-gen generates sane code.

cc @deads2k for #20573

### CODE REVIEW

**Comment by caesarxuchao** [comment]
> because golang doesn't allow dot in variable/function names, so when the group name is used as part of variable/function name, client-gen extracts the part before the first dot (e.g., authentication).

@lavalamp, because apart from `authentication.k8s.io`, perhaps there will be group `authentication.openshift.io` in the future, perhaps I should just remove the dots rather than extracting `authentication` from the full group name?

File: cmd/libs/go2idl/client-gen/main.go [hunk 1]
```diff
@@ -92,12 +92,12 @@ func main() {
 
 	if *test {
 		arguments.InputDirs = append(dependencies, []string{
-			"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup",
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/main.go]: I'll simplify the test setup in another PR.

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this, issue #23350

**Comment by lavalamp** [comment]
> because apart from authentication.k8s.io, perhaps there will be group authentication.openshift.io in the future, perhaps I should just remove the dots rather than extracting authentication from the full group name?

I think not? People will typically be importing a particular version from within, anyway.

Unless it causes problems within the autogenerator itself. I guess I vote to fix it when it happens to us :)

File: cmd/libs/go2idl/client-gen/testoutput/clientset_generated/test_internalclientset/typed/testgroup.k8s.io/unversioned/testgroup_test.go [hunk 2]
```diff
@@ -21,9 +21,9 @@ import (
 	"net/url"
 	"testing"
 
-	"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup"
-	_ "k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup/install"
-	. "k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testoutput/clientset_generated/test_internalclientset/typed/testgroup/unversioned"
+	testgroup "k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io"
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/testoutput/clientset_generated/test_internalclientset/typed/testgroup.k8s.io/unversioned/testgroup_test.go]: Why isn't this one `testgroup_k8s_io` too?

Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/testoutput/clientset_generated/test_internalclientset/typed/testgroup.k8s.io/unversioned/testgroup_test.go]: This is a manual change. I believe the other places that uses `_` to concatenate elements are generated code. I can make the change here if you think that's necessary.

Comment (krousey) [inline cmd/libs/go2idl/client-gen/testoutput/clientset_generated/test_internalclientset/typed/testgroup.k8s.io/unversioned/testgroup_test.go]: No it's fine.

**Comment by krousey** [comment]
LGTM. Squash however you like and add the label.

**Comment by caesarxuchao** [comment]
Thanks. Squashed and label applied.

**Comment by caesarxuchao** [comment]
Adding a priority label as it's blocking other PRs.

**Pull request merged**